### DISCUSSION TITLE
The syntax of pattern matching

### DISCUSSION THREAD

**Discussion by kyouko-taiga** [hylo-lang/hylo#1195] [Language design]
Our implementation of pattern matching has significantly drifted from our specification so I think it may be good to have a look at the current state of things. Today, pattern matching looks like [Swift](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/patterns/) but with a couple of notable syntactic differences.

First, don't use the `is`, `as` and `as?` operators; instead we use `:` to match over types. For example:

```swift
// In Swift
let x: Any = ...
switch x {
  case let y as T:
    print("a T equal to \(y)")
  case is U:
    print("an arbitrary U") }
  default:
    print("something else")
}

// In Hylo
let x: Any = ...
match x {
  case let y : T
    { print("a T equal to \(y)") } 
  case _: U
    { print(y) } 
  case _
    { print("something else") }
}
```

The reason for having `match` instead of `switch` is kind of cosmetic but my sense is that Swift picked the latter to make simple switch statements (e.g., over integers) look familiar to a C-family audience. IMO, this construction is going much farther than C's switch and relates more to pattern matching as popularized by other languages like [Scala](https://docs.scala-lang.org/tour/pattern-matching.html) or [Ocaml](https://v2.ocaml.org/manual/patterns.html).

I find the use of `as` and `as?` confusing in Swift. The latter must be used in `if` statements and the former in all other forms of matching. This rule seems questionably (pun intended) motivated to me. In contrast, I find `:` more concise and using it consistently feels less confusing.

A second difference is that we don't use `case` to introduce patterns in `if` statement/expression or a `for` loop:

```swift
// In Swift
let x: [Any] = ...
for case let y as T in x { print(y) }

// In Hylo
let x: Array<Any> = ...
for let y: T in x { print(y) }
```

The absence of `case` makes our statements/expressions more concise without removing clarity. The same couldn't be said in `match`, however, because patterns may become difficult to identify in that expression if you (as a human) quickly parse it. To understand what I mean, let's examine the following expression:

```swift
match x { _: T { a } _ { b } }
```

The structure of our AST currently makes it impossible to introduce names at arbitrary depths in structural pattern. For example, this program is syntactically illegal:

```swift
let x = (a: 1, b: 2)
if (a: 1, b: let y) = x { print(y) }
```

I am not sure it is something we actually want, though, so when I discovered the limitation I left it there. We can often rewrite this kind of code using a side condition on the match, or even write another match in the success clause. However, these workarounds are more verbose and so it is quite possible that we may want to lift the limitation.

A last notable change is that optional binding in Hylo must use an optional pattern rather than Swift's more concise syntax. This difference makes optionals behave more consistently w.r.t. to other union types because `let y = x` is irrefutable and doesn't narrow `x`'s type in Hylo.

```swift
// In Swift
let x = Optional(1)
if let y = x { print(y) }

// In Hylo
let x = Optional(1)
if let y? = x { print(y) }
```

_Note: a variant of the Hylo syntax does [exist in Swift](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/patterns/#Optional-Pattern)._

## Ideas for extensions

It would be nice if we could pattern match on non-structural types. Swift has some support of [custom patterns](https://www.swiftbysundell.com/articles/defining-custom-patterns-in-swift/) but I find it lacking. AFAIK, there's no way to "bind" values with a custom pattern; we can only match existing values.

### Custom patterns

One way to support more powerful custom patterns would be to define a trait that requires a projection:

```swift
public trait Inspectable {
  type Scrutinee
  property scrutinee: Scrutinee { let }
}

type Point: Regular, Inspectable {
  public var x: Int
  public var y: Int
  public memberwise init

  public typealias Scrutinee = { x: Int, y: Int }
  public property Scrutinee: { (x: x, y: y) }
}

public fun main() {
  let p = Point(x: 4, y: 2)
  match p {
    case (x: _, y: 2) { ... }
    case _ { ... }
  }
}
```

Of course, we could have different effect requirements on `scrutinee` to support `inout` patterns and `sink` patterns too.

### List patterns

I wrote this extension in the utilities of `hc`, which I use all the time.

```swift
extension Collection {
  /// The first element of `self` and its suffix after its first index or `nil` if `self` is empty.
  public var headAndTail: (head: Element, tail: SubSequence)? {
    if isEmpty { return nil }
    return (head: self[startIndex], tail: self[index(after: startIndex)...])
  }
}
```

You'll notice that this method relates to a non-empty list pattern in [Haskell](https://en.wikibooks.org/wiki/Haskell/Pattern_matching). It would be nice if we could use it in pattern matching.

I think we could with the following trait:

```swift
public trait InspectableAsCollection {
  type Element
  type SubCollection
  property collection_scrutinee: Optional<remote let {head: Element, tail: SubCollection}> { let }
}

// Or perhaps Collection : InspectableAsCollection ?
public conformance Array: InspectableAsCollection { ... }

public fun main() {
  let a = Array([1, 2, 3])
  match a {
    case let (h +... t) where t.is_empty()
      { /* matches a collection of one element */ }
    case let (h +... t)
      { /* matches the head and tail of a non-empty collection */ }
    case _
      { /* collection is empty if we're here */ }
  }
}
```

**Comment by dabrahams** [comment]
I'm just noticing in two of your Hylo examples:

```swift
for let y: T in x { print(y) }
```

and 

```swift
if (a: 1, b: let y) = x { print(y) }
```

The inconsistency of the placement of `let` is surprising and the apparent use of `y` as a type name in the 2nd example is confusing.  Or maybe I'm not reading the 2nd example correctly?  Oh, I see, I'm not.  Fact remains that the possiblity of these two syntaxes being legal is potentially confusing, so I'm leaving that part there as a marker.

OK, you say the 2nd example is “introducing names at arbitrary depths in structural pattern.”  The introduced name is `y` here, right?  `y` is at depth 1, where at depth 0 you have the entire pattern.  So it sounds like you're saying you can't do a destructuring bind as part of a pattern match and instead must separately match and bind, e.g.:

```swift
if (a: 1, b: _) = x { let y = ??; print(y) }
```

Actually I don't know what I'd even write for `??`.  I can't write `x.b` in the general case because x could have union type, so `.b` is not necessarily statically knowable (and I hope we're not silently changing the type of `x` inside the `if` body based on the pattern match—that was proposed several times for Swift but I think it's confusing).

Anyway, I'm surely not interpreting your words correctly because not being able to pattern match and bind to a part in one expression seems awfully and needlessly inconvenient.

**Reply by kyouko-taiga** [reply]
> Oh, I see, I'm not. Fact remains that the possiblity of these two syntaxes being legal is potentially confusing, so I'm leaving that part there as a marker.


I'm not arguing against the confusion but note that I didn't invent any syntax here:

```swift
// In Swift:
let x = (a: 1, b: 2)
if case (a: let n, b; let m) = x { print(n + m) }
```

One difference is that Swift lets you ignore tuple labels, which in many case makes sense.

>  The introduced name is y here, right?  y is at depth 1, where at depth 0 you have the entire pattern.

Correct.

> So it sounds like you're saying you can't do a destructuring bind as part of a pattern match and instead must separately match and bind, e.g.:

I meant that we can match the whole thing and then use narrowing conditions to do what the value match would have done:

```
if let z = x, z.a == 1 { let y = z.b; print(y) }
```

_Well, it's silly to write `let z = x` in a `if` condition but we can imagine having this particular case in a `match` statement._

I think this constraint only adds verbosity without significantly reducing expressiveness. Value matches are equivalent to Boolean tests in a where clause. There are probably some crazy setups we could imagine using different binding introducers, e.g., `(a: let x, b: sink y: c: 3)` but I think we can at least try to see if we can deal without that.

> and I hope we're not silently changing the type of x inside the if body based on the pattern match—that was proposed several times for Swift but I think it's confusing

Sadly the nth time worked in Swift. We can write `if x { print(x) }` when `x` is an optional and it will "change type" in the then clause. I guess it's only a matter of time before someone write a SE proposal for `if x is Int { x + 1 }`.

For Hylo I (or perhaps _we_) want to maintain that all value bindings are introduced by `let`, `inout`, `sink`, or `set`. That is why I'm also not a fan of another common syntax for binding in patterns, e.g. Scala `x match case (n: Int, m: Int) => print(n + m)`.

**Reply by dabrahams** [reply]
> I think this constraint only adds verbosity without significantly reducing expressiveness.

OK… but is it better?  It changes something declarative into something imperative, IMO.  Is that not a change in expressiveness?

> I (or perhaps we) want to maintain…

_We_ do.

**Reply by dabrahams** [reply]
Aside: Is this really how you want to spell computed properties, or am I mis-guessing the meaning of this construct?

> `subscript scrutinee: T { let }`

I do understand the desire to draw the connection between computed properties and subscripts but IMO declaring what acts like a property this way is too weird for the average programmer to accept.  It feels like forcing something distasteful on the user because “it's good for you.”

**Reply by dabrahams** [reply]
Is this what you mean?

```swift
/// Types with a non-nominal representation that can be pattern-matched.
public trait Structural {
  /// The non-nominal representation.
  type Structure
  
  /// The non-nominal representation.
  subscript structure: Structure { let }
}

type Point: Regular, Structural {
  public var x: Int
  public var y: Int
  public memberwise init
  public type Structure = (x: Int, y: Int)
  public subscript structure: { (x: x, y: y) }
}
```

**Reply by kyouko-taiga** [reply]
> OK… but is it better? It changes something declarative into something imperative, IMO. Is that not a change in expressiveness?

I don't have a strong opinion as to whether or not the limitation should stay. I don't think I have many examples of cases where writing a pattern of the form `(1, let x)` would have helped me so I'd say the limitation isn't harmful and makes the language simpler. But I can easily be convinced otherwise.

**Reply by dabrahams** [reply]
Suggestion: 

```swift
extension Collection {
  /// The first element of `self` and the rest of its elements, or `nil` if `self` is empty.
  public var headAndTail: (head: Element, tail: SubSequence)? {
    first.map { (head: $0, tail: dropFirst()) }
  }
}
```

**Reply by kyouko-taiga** [reply]
> Is this really how you want to spell computed properties, or am I mis-guessing the meaning of this construct?

It should have been `property`, sorry. It's clearly a computed property and it should have the same syntax and semantics.

**Reply by dabrahams** [reply]
> But I can easily be convinced otherwise.

I'm glad you have no strong position and don't feel the need to convince you until real use-cases present contrary evidence.

**Reply by dabrahams** [reply]
@kyouko-taiga https://github.com/orgs/hylo-lang/discussions/1195#discussioncomment-8822064 ?
I didn't see an associated type declaration in your code but I assume you need one, unless I'm misunderstanding the whole idea.

**Reply by kyouko-taiga** [reply]
Yes, another mistake ...
I'm updating the original post.

**Comment by kyouko-taiga** [comment]
See also: https://docs.scala-lang.org/scala3/reference/changed-features/pattern-matching.html

**Comment by tothambrus11** [comment]
> ```swift
> // In Swift
> let x: [Any] = ...
> for case let y as T in x { print(y) }
> 
> // In Hylo
> let x: Array<Any> = ...
> for let y: T in x { print(y) }
> ```
I find the second syntax a bit confusing. Usually, we are allowed to write the type of the binding after its declaration like `let myVar: Int = getMyNumber()` to constrain its type and make sure that the right hand side matches it. The problem with this syntax is that it might actually work for some time, but when the e.g. library writer changes `x` to be a collection of `CaillerChocolate`s, the above code would silently turn into a filter.

@kyouko-taiga earlier mentioned a special where syntax for for loops, I think that would be a better alternative:
```hylo
for let e in x where e : T { 
  .... 
}
```
...Or some generic algorithm that can utilize pattern matching as predicates. For this, I think the `is` keyword is actually useful from Swift, as it would look too ambiguous to write without a keyword:
```hylo
let filtered = x.filter({ $0 is T })

// or

for let t in x.filter({ $0 is T }) {
  print(x)
}
```

**Comment by kyouko-taiga** [comment]
I'm reviving this thread because I lately I thought about custom extractors. I think it is important that we can express refutable patterns, which is not possible in general with my initial proposal.

To illustrate, imagine we'd like to use pointer tagging to optimize an enum like this:

```hylo
enum Data {
  case inline(size: UInt8, data: UInt8[7])
  case out_of_line(data: DynamicBuffer<UInt8>)
}
```

with a structure like this, while keeping the ability to apply pattern matching on the two (notional) cases:

```hylo
struct Data {
  var representation: UInt64
}
```

The essential operations we have to do to pattern match on the struct are the same as those we'd have to do on the enum:

1. look at the discriminator
2. if that's the one we're looking for, access the payload and project its parts
3. otherwise, move to the next pattern

We can do most of this work with a subscript. The only issue is to tell whether or not something did match. Intuitively, we can address this issue with the following function:

```hylo
fun out_of_line<E>(
  scrutinee: Data,
  action: [E](BufferPointer<UInt8>) sink -> Void
) -> Bool { ... }
```

If matching succeeds, then `out_of_line` applies `action` and returns `true`. Otherwise, it returns `false`. This "protocol" could be used to implement pattern matching. We could even use method bundles to support different binders for extracted values. Then, with some sugars, a function like the one above could serve as a custom extractor.

Interestingly, I think this approach also addresses the issue of projecting optional values from collections. This idea has been floated by @dabrahams I believe. For example, we could write `Array.first[:]` like this:

```hylo
extension <T> T[] {
  fun first<T, E>(action: [E](T) sink -> Void) -> Bool { ... }
}

public fun main() {
  let xs: Array<_> = [1, 2, 3]
  if .first(let x) = xs { ... }
}
```

It may be surprising to use pattern matching only to extract parts of a value rather than the whole, though ...

**Reply by dabrahams** [reply]
I'm having trouble reading the last example.  I'll assume

```scala
extension <T> T[] {
```
is the same as
```scala
extension <T> Array<T> {
```

and that we get to use leading dot sugar in the `if` because `xs` is an `Array`.  But `first` was declared as a function and you seem to be invoking it as a subscript?  And what lets you write `let x` there?  Is this all special sugar that I haven't seen before?

**Reply by kyouko-taiga** [reply]
> I'll assume [...] is the same as [...]

Yes, `T[]` is (tentatively) sugar for `Array<T>`.

> and that we get to use leading dot sugar in the if because xs is an Array. 

We can use the leading dot in any pattern position, not specifically because `xs` is an array. IIRC we agreed that there should be no need for a `case` keyword in conditions.

Conditions have this syntax (selected rules):

```
condition ::=
  | condition-item (',' condition)?

condition-item ::=
  | expression
  | pattern '=' expression

pattern ::=
  | identifier
  | 'let' pattern (':' expression)?
  | '.' identifier '(' pattern-argument-list ')'

pattern-argument-list ::=
  | pattern-argument (',' pattern-argument-list)? ','?

pattern-argument ::=
  | label? pattern
```

> But first was declared as a function and you seem to be invoking it as a subscript?

I should have used parentheses (now fixed in the original post). I initially declared `first` as a subscript but changed it to a function when I realized a subscript was not necessary. However, if we go with the idea of having optional subscripts we would probably require subscript syntax at matching site.

> And what lets you write let x there?

That's because we're parsing an argument pattern (see above grammar excerpt).

**Reply by dabrahams** [reply]
> We can use the leading dot in any pattern position

I understand that.  The fact that it finds `.first` is because `xs` is an `Array` though, right?

**Reply by kyouko-taiga** [reply]
> The fact that it finds .first is because xs is an Array though, right?

Correct.

**Comment by dabrahams** [comment]
In a recent meeting we discussed a way to connect subscripts and pattern matching but I don't remember the details.  Would be good if it could be captured here.

**Reply by kyouko-taiga** [reply]
We said that an extractor written as a higher-order function like this

```
extension <T> Array<T> {
  fun first<E>(action: inout [E](T) inout -> Void) -> Bool {
    if self.empty() { return false }
    &action(self[0])
    return true
  }
} 
```

could be written as an optional subscript like that

```
extension <T> Array<T> {
  subscript first?() -> T {
    if self.empty() { return }
    yield self[0]
  }
} 
```

We also mentioned there was an issue for subscript projecting more than one value, as we can't capture remote parts into tuples. But we said yielding a tuple in an optional subscript could be blessed with special semantics.

**Reply by dabrahams** [reply]
MO we could capture remote parts in tuples and non resilient enums because like closures they too are structural

**Comment by tothambrus11** [comment]
I really like Swift's `if case`, but sometimes I come across problems when I really just want to get a boolean expression on whether the case matched or not:
```swift
    /// Returns `true` iff `self` denotes the body of at type declaration.
    var isTypeBody: Bool {
      if case .typeBody = self { return true } else { return false }
    }

    /// Returns `true` iff `self` denotes the body of a trait declaration.
    var isTraitBody: Bool {
      if case .typeBody(let t) = self { 
        t == TraitDeclaration.self
      } else { 
        false
      }
    }
 ```
 
 So it would be nice to have some operator or language construct which can be used similarly as the `is` keyword, just for matching enums. This renders any associated values useless, since there is no scope in which they would be definitely present. So I would disallow binding variables in an extractor pattern.

**Reply by dabrahams** [reply]
we already decided that `if self.typeBody != nil` should be made to work, IIRC

**Reply by dabrahams** [reply]
Err, I suppose that might not be quite the right API because we don't have `Optional<remote T>`.  So maybe the API is:

```
let x: Optional<Q> = self.typeBody[(payload: P)->Q { ... return someQ }]
```

or something, and then maybe something like:

```
if self.typeBody[(_){}] != nil
```
works for your case, modulo me getting the syntax wrong.

**Reply by tothambrus11** [reply]
Or we could have `self.typeBody` be a boolean computed property. One can always get the payload by pattern matching in a conditional. (It appears to me that having `.typeBody`  be a a map-like higher order function/subscript would make things more complicated while having similar expressiveness as a pattern matching if expression.)

**Reply by kyouko-taiga** [reply]
I think we can simply have an operator that tests whether or not some pattern matches.

```
let b = self ~= .type_body(_)
```

We've be reserving `~=` as a special operator so that we can parse the rhs is a pattern, and then we could use any pattern we like.

**Reply by tothambrus11** [reply]
Testing for arbitrary patterns would be very nice! `~=` to me looks a bit like `+=`, reminding me of mutation. If we have a `match` keyword for pattern matching as conditionals, a natural extension would be `x matches y` to test a pattern.

**Reply by kyouko-taiga** [reply]
Not to say that using a keyword is a bad idea altogether but I fine the rationale a little weak. `-=`,`*=`, `|=` also look like `+=` if you squint your eyes.

`~=` has precedent in Swift, and it works there. As for the "natural extension", while we can make it a soft keyword, `matches` is long and  I would not like to have to spell it every time I want to use pattern matching.

```
if let x? = foo { print(x) }
```

looks better than

```
if foo matches .some(let x) { print(x) }
```

I considered `is` but I think we shouldn't overload its meaning since we already use it as sugar for conformance witnesses.

**Reply by tothambrus11** [reply]
> -=,*=, |= also look like += if you squint your eyes.

Yes, that was exactly my feeling. `~=` looked a bit like as it belonged to the "perform operation and apply" family of operators. 

I get the point made by your examples, I just want to point out they are slightly different use cases. I'm onboard with having the first one already, which is used for binding a variable conditionally based on whether the pattern matched. `matches`/`is` would just evaluate to a boolean without binding any variables.

I think we would use `matches` relatively rarely (given that no other language I know has this) and most often we would use either `match` or `if let`. So I think the long name wouldn't hurt much, while being easy to remember, discover and read.

**Reply by kyouko-taiga** [reply]
You can make the same reasoning about `~=` looking like `==`, `!=`, `<=`, etc. and therefore conclude that it belongs to the "compare" family of operators.

**Reply by tothambrus11** [reply]
You are right, fair enough.

**Reply by dabrahams** [reply]
You can use `=~` to avoid the “it looks like an in-place operator” problem somewhat.

**Comment by tothambrus11** [comment]
> ```hylo
> for let y: T in x { print(y) }
> ```
In the pattern matching filtering `for` loop example, it is unclear to me what distinguishes it from a regular iteration with explicit type annotation on the binding. IMO having a wrong type annotation should result in a compiler error rather than silently turning my loop into a filtering loop.

We should include something extra if we want a match, maybe something along the lines of `for let y?: T`. I'm a bit reluctant to this, we should avoid a lot of different custom syntax in different contexts. I always need to look up what's the syntax for Swift's `if case` and pattern matching (might be an issue with my memory, but still something I'd like to avoid for Hylo)
